subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java

```diff
@@ -79,66 +79,54 @@ public boolean isFetchingMetadataCheap(ComponentIdentifier identifier) {
     private void resolveModule(ModuleComponentIdentifier identifier, ComponentOverrideMetadata componentOverrideMetadata, BuildableComponentResolveResult result) {
         LOGGER.debug("Attempting to resolve component for {} using repositories {}", identifier, repositoryNames);
 
-        List<Throwable> errors = new ArrayList<Throwable>();
-
         List<ComponentMetaDataResolveState> resolveStates = new ArrayList<ComponentMetaDataResolveState>();
         for (ModuleComponentRepository repository : repositories) {
             resolveStates.add(new ComponentMetaDataResolveState(identifier, componentOverrideMetadata, repository, versionedComponentChooser));
         }
 
-        final RepositoryChainModuleResolution latestResolved = findBestMatch(resolveStates, errors);
-        if (latestResolved != null) {
-            LOGGER.debug("Using {} from {}", latestResolved.module.getId(), latestResolved.repository);
-            for (Throwable error : errors) {
-                LOGGER.debug("Discarding resolve failure.", error);
-            }
-
-            result.resolved(metaDataFactory.transform(latestResolved));
-            return;
-        }
-        if (!errors.isEmpty()) {
-            result.failed(new ModuleVersionResolveException(identifier, errors));
-        } else {
-            for (ComponentMetaDataResolveState resolveState : resolveStates) {
-                resolveState.applyTo(result);
+        try {
+            final RepositoryChainModuleResolution latestResolved = findBestMatch(resolveStates);
+            if (latestResolved == null) {
+                for (ComponentMetaDataResolveState resolveState : resolveStates) {
+                    resolveState.applyTo(result);
+                }
+                result.notFound(identifier);
+            } else {
+                LOGGER.debug("Using {} from {}", latestResolved.module.getId(), latestResolved.repository);
+                result.resolved(metaDataFactory.transform(latestResolved));
             }
-            result.notFound(identifier);
+        } catch (Throwable error) {
+            result.failed(new ModuleVersionResolveException(identifier, error));
         }
     }
 
-    private RepositoryChainModuleResolution findBestMatch(List<ComponentMetaDataResolveState> resolveStates, Collection<Throwable> failures) {
+    private RepositoryChainModuleResolution findBestMatch(List<ComponentMetaDataResolveState> resolveStates) throws Throwable {
         LinkedList<ComponentMetaDataResolveState> queue = new LinkedList<ComponentMetaDataResolveState>();
         queue.addAll(resolveStates);
 
         LinkedList<ComponentMetaDataResolveState> missing = new LinkedList<ComponentMetaDataResolveState>();
 
         // A first pass to do local resolves only
-        RepositoryChainModuleResolution best = findBestMatch(queue, failures, missing);
+        RepositoryChainModuleResolution best = findBestMatch(queue, missing);
```

Comment:
❌ What if the local repo has a problem (e.g. a broken maven-metadata.xml), but the remote one is fine?

Reply:
Can you check if we have a test for this behavior? It is ok if it fails, but it should be like this in the current version already. We need to make sure that we do not change the behavior.

Reply:
@jjohannes Done here:https://github.com/gradle/gradle/blob/21712f94de5cd0096b0c55ece8bac8cd7bd68af8/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenLocalRepoResolveIntegrationTest.groovy#L127

I'm not sure if this is correct, could a repository have both local access and remote access at the same time?

Reply:
The test is ok. We always first try the local access and then the remote access (if local is missing) as far as I understand. On error in local, we exit directly (without trying remote).
`ComponentMetaDataResolveState` always first tries local access and if it fails with an error, the complete _resolve_ operation fails. The test above however fails before we even get there.

Another test that exercises this code path would be good as well. For that situation, the local metadata needs to be corrupt. We could do a similar test as above:

```
      given:
        def corruptLocalPom = m2.mavenRepo().rootDir.file('group/projectA/1.2/projectA-1.2.pom')
        corruptLocalPom << "invalid content"

        if (remoteStatus == 'valid') {
            def remoteRepo = maven("remote")
            remoteRepo.module('group', 'projectA', '1.2').publish()

            buildFile << """
            repositories{
                maven { url "${remoteRepo.uri}" }
            }
            """
        }

        when:
        runAndFail 'retrieve'

        then:
        failure.assertHasCause("Could not parse POM $corruptLocalPom.absolutePath")

        where:
        remoteStatus << ['valid', 'invalid']
```

:x: In fact, it looks like this test fails without the change. We might need to rethink the change with the error handling. 🙁

Reply:
Looks like it's not mocking a failure in local access here because `m2.mavenRepo().rootDir.file('group/projectA/1.2/projectA-1.2.pom')` is `mavenLocal()`, so actually there're 2 repos here: `mavenLocal()` and `remoteRepo`.

This test is testing fall through next repository.

Reply:
Yes you are right. This change is expected. Sorry for the confusion.